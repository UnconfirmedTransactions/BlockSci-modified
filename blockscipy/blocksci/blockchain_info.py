import requests
import pandas as pd

class BlockchainInfoData(object):
    def __init__(self, api_key, cache_blocks=False, cache_txs=False):
        self.base_url = 'https://blockchain.info/'
        self.api_key = api_key
        self.cache_blocks = cache_blocks
        self.cache_txs = cache_txs
        self.block_cache = {}
        self.tx_cache = {}
        
    def block_data(self, block_hash):
        if block_hash in self.block_cache:
            return self.block_cache[block_hash]       
        
        url = self.base_url + 'rawblock/' + block_hash + "?api_code=" + self.api_key
        r = requests.get(url)
        r.raise_for_status()
        
        data = r.json()
        if self.cache_blocks:
            self.block_cache[block_hash] = data
        return data
    
    def tx_data(self, tx_hash):
        if tx_hash in self.tx_cache:
            return self.tx_cache[tx_hash]

        url = self.base_url + 'rawtx/' + tx_hash + "?api_code=" + self.api_key
        r = requests.get(url)
        r.raise_for_status()
        
        data = r.json()
        if self.cache_txs:
            self.tx_cache[tx_hash] = data
        return data
    
    def block_arrival_time(self, block_hash):
        data = self.block_data(block_hash)
        
        if "received_time" in data:
            return pd.to_datetime(data["received_time"], unit='s')
        else:
            return None
    
    def tx_arrival_times(self, block_hash):
        data = self.block_data(block_hash)
        arrival_times = {}
        for tx in data["tx"]:
            arrival_times[tx["hash"]] = pd.to_datetime(tx["time"], unit='s')
        return arrival_times
    
    def tx_arrival_time(self, tx_hash):
        data = self.tx_data(tx_hash)
        if "time" in data:
            return pd.to_datetime(data["time"], unit="s")
        else:
            return None
    
    def block_relayed_by(self, block_hash):
        data = self.block_data(block_hash)
        if "relayed_by" in data:
            return data["relayed_by"]
        else:
            return None