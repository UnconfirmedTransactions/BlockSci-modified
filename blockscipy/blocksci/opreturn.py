# -*- coding: utf-8 -*-
"""BlockSci Module

Label op return data as related to a specific application
"""

import binascii
import re
import Crypto.Cipher.ARC4

OP_RETURN_SERVICES = {
    "OA": "Open Assets", "id": "Blockstack",
    "S1": "Stampery", "S2": "Stampery",
    "S3": "Stampery", "S4": "Stampery", "S5": "Stampery",
    "Factom!!": "Factom", "FACTOM00": "Factom", "Fa": "Factom", "FA": "Factom",
    "ORIGMY": "OriginalMy",
    "ProveBit": "ProveBit",
    "UNicDC": "University of Nicosia",
    "DOCPROOF": "Proof of Existence",
    "omni": "Omni Layer", "ASCRIBESPOOL": "Ascribe",
    "CC": "Colu", "MG": "Monegraph",
    "SB.D": "SB.D", "BITPROOF": "Bitproof",
    "KC": "KC", "BS": "BlockSign", "OC": "UnknownOC",
    "CryptoTests-": "Crypto Copyright", "CryptoProof-": "Crypto Copyright",
    "LaPreuve": "LaPreuve",
    "RMBd": "Remembr", "RMBe": "Remembr",
    "Mined by 1hash.com": "1hash",
    "FluxST": "FluxST", "CP110400": "CP1", "KMD": "Komodo", "OKT": "OKT"
}

BYTE_PREFIXES = {
    b'g\x01\xdd6\x15&]+': "UnknownBytePrefix",
    binascii.unhexlify(b'5888'): "Blockstack",
    binascii.unhexlify(b'5808'): "Blockstack",
    binascii.unhexlify(b'455720'): "Eternity Wall",
    binascii.unhexlify(b'53504b'): "Coinspark",
    binascii.unhexlify(b'4f43'): "Openchain",
    b'STAMPD##': "stampd",
    b'CNTRPRTY': "Counterparty"
}
EXACT_STRING_MATCHES = [
    "http://www.blockcypher.com/",
    "http://www.blockcypher.com",
    "DS",
    "503: Bitcoin over capacity!",
    "XX",
    "XY",
    "XW",
    "SS"
]
EXACT_BYTE_MATCHES = [
    b'\xb3\x87\xfc~\xf6K3\xcf\x01\x82\xec\xf8\xea\xb2\x065\x8cz\xc7\xb5\xa9\xd1$\x1a\x11\xd4Sb\xda\x9af\xa2\xb9l\xb6\xb6\x9b\xd7\xa6a',
    b"\xe3\xb0\xc4B\x98\xfc\x1c\x14\x9a\xfb\xf4\xc8\x99o\xb9$'\xaeA\xe4d\x9b\x93L\xa4\x95\x99\x1bxR\xb8U",
    b'\xfc\xf4=\xa2D\x99GZ\x96\x00#1\xb8\x97\xc8\xf3#+(\xe5z\xfd\xcb?}y\x91\xbfv\xc8\xf0\xb6\x05\x81\xa2\x89\x9d\x80\x9f\xa9',
    b'*K\x94V\x827C\xfe\xc9\xc0\x7fv:\x9f\x0c\xa3@~\xba\xb2\xfa\xca\xe4\x0e\xbbm\x83\xc8\xf5k\x0b\xbf\x94\xcbn\xe7\xb0\x9a$\r'
]

# address_matches = {
#     Address(155316552, address_type.pubkeyhash):"Chainpoint", # 12m3EH1kCFWc9WvHmtLoby17xuYnSdNVYR 
#     Address(89427334, address_type.pubkeyhash):"Base64DataAddress1",
#     Address(156665979, address_type.pubkeyhash):"Base64DataAddress2",
#     Address(182075165, address_type.pubkeyhash):"Base64DataAddress3",
#     Address(156599323, address_type.pubkeyhash):"Base64DataAddress4",
#     Address(203933727, address_type.pubkeyhash):"Komodo",
#     Address(203831434, address_type.pubkey):"Komodo",
#     Address(127559021, address_type.pubkeyhash): "UnknownAddress1",
#     Address(16179390, address_type.scripthash):"SpamAddress1",
#     Address(16163151, address_type.scripthash):"SpamAddress2",
#     Address(89403362, address_type.pubkeyhash):"SpamAddress3",
#     Address(155249896, address_type.pubkeyhash):"SpamAddress4",
#     Address(77604537, address_type.pubkeyhash):"SpamAddress5",
#     Address(127504441, address_type.pubkeyhash):"SpamAddress5"
# }


def label_application(tx):
    if tx.op_return is None:
        return None
    data = tx.op_return.address.data

    if not data:
        return "Empty"

    decoded = data.decode("ascii", "replace")
    strings = re.findall("[a-zA-Z0-9.://! ]+", decoded)

    if decoded in EXACT_STRING_MATCHES:
        return decoded
    elif data in EXACT_BYTE_MATCHES:
        return data
    elif strings and strings[0].startswith(tuple(OP_RETURN_SERVICES.keys())):
        for prefix in OP_RETURN_SERVICES:
            if strings[0].startswith(prefix):
                return OP_RETURN_SERVICES[prefix]
    elif data.startswith(tuple(BYTE_PREFIXES.keys())):
        for prefix in BYTE_PREFIXES:
            if data.startswith(prefix):
                return BYTE_PREFIXES[prefix]
    # elif len(set(txout.address for txout in tx.outs).intersection(set(address_matches.keys()))) > 0:
    #     for txout in tx.outs:
    #         if txout.address in address_matches:
    #             return address_matches[txout.address]
    elif tx.ins:
        first_vin_txid = binascii.unhexlify(str(tx.ins[0].spent_tx.hash))
        decoded = Crypto.Cipher.ARC4.new(first_vin_txid).decrypt(data)
        if decoded.startswith(b'CNTRPRTY'):
            return "Counterparty"
    return "Unknown"
