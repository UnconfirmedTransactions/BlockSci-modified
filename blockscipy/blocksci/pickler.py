from ._blocksci import *
import pickle
import sqlite3
from collections import namedtuple

# Simple class representing a record in our database.
MemoRecord = namedtuple("MemoRecord", "key, task")

class Pickler(pickle.Pickler):
    """Custom Pickler for BlockSci objects"""

    def persistent_id(self, obj):
        # Instead of pickling MemoRecord as a regular class instance, we emit a
        # persistent ID.
        if isinstance(obj, Block):
            # Here, our persistent ID is simply a tuple, containing a tag and a
            # key, which refers to a specific record in the database.
            return ("Block", obj.height)
        elif isinstance(obj, Tx):
            return ("Tx", obj.index)
        elif isinstance(obj, Output):
            return ("Output", (obj.tx_index, obj.index))
        elif isinstance(obj, Input):
            return ("Input", (obj.tx_index, obj.index))
        elif isinstance(obj, Address):
            return ("Address", (obj.address_num, obj.type))
        elif isinstance(obj, EquivAddress):
            sample_addr = next(iter(obj))
            return ("EquivAddress", (sample_addr.address_num, sample_addr.type, obj.is_script_equiv))
        else:
            # If obj does not have a persistent ID, return None. This means obj
            # needs to be pickled as usual.
            return None


class Unpickler(pickle.Unpickler):
    """Custom Unpickler for BlockSci objects"""

    def __init__(self, file, chain):
        super().__init__(file)
        self.chain = chain

    def persistent_load(self, pid):
        # This method is invoked whenever a persistent ID is encountered.
        # Here, pid is the tuple returned by DBPickler.
        type_tag, key_id = pid
        if type_tag == "Block":
            return self.chain[key_id]
        elif type_tag == "Tx":
            return self.chain.tx_with_index(key_id)
        elif type_tag == "Output":
            return self.chain.tx_with_index(key_id[0]).outputs[key_id[1]]
        elif type_tag == "Input":
            return self.chain.tx_with_index(key_id[0]).inputs[key_id[1]]
        elif type_tag == "Address":
            return self.chain.address_from_index(key_id[0], key_id[1])
        elif type_tag == "EquivAddress":
            addr = self.chain.address_from_index(key_id[0], key_id[1])
            return addr.equiv(key_id[2])
        else:
            # Always raises an error if you cannot return the correct object.
            # Otherwise, the unpickler will think None is the object referenced
            # by the persistent ID.
            raise pickle.UnpicklingError("unsupported persistent object")
