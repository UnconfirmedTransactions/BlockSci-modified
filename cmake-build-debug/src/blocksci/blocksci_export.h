
#ifndef BLOCKSCI_EXPORT_H
#define BLOCKSCI_EXPORT_H

#ifdef BLOCKSCI_STATIC_DEFINE
#  define BLOCKSCI_EXPORT
#  define BLOCKSCI_NO_EXPORT
#else
#  ifndef BLOCKSCI_EXPORT
#    ifdef blocksci_EXPORTS
        /* We are building this library */
#      define BLOCKSCI_EXPORT __attribute__((visibility("default")))
#    else
        /* We are using this library */
#      define BLOCKSCI_EXPORT __attribute__((visibility("default")))
#    endif
#  endif

#  ifndef BLOCKSCI_NO_EXPORT
#    define BLOCKSCI_NO_EXPORT __attribute__((visibility("hidden")))
#  endif
#endif

#ifndef BLOCKSCI_DEPRECATED
#  define BLOCKSCI_DEPRECATED __attribute__ ((__deprecated__))
#endif

#ifndef BLOCKSCI_DEPRECATED_EXPORT
#  define BLOCKSCI_DEPRECATED_EXPORT BLOCKSCI_EXPORT BLOCKSCI_DEPRECATED
#endif

#ifndef BLOCKSCI_DEPRECATED_NO_EXPORT
#  define BLOCKSCI_DEPRECATED_NO_EXPORT BLOCKSCI_NO_EXPORT BLOCKSCI_DEPRECATED
#endif

#if 0 /* DEFINE_NO_DEPRECATED */
#  ifndef BLOCKSCI_NO_DEPRECATED
#    define BLOCKSCI_NO_DEPRECATED
#  endif
#endif

#endif /* BLOCKSCI_EXPORT_H */
