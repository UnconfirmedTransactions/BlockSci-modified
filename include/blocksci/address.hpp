//
//  address.hpp
//  blocksci
//
//  Created by Harry Kalodner on 12/12/17.
//
//

#ifndef blocksci_address_group_header
#define blocksci_address_group_header

#include <blocksci/address/address.hpp>
#include <blocksci/address/equiv_address.hpp>

#endif /* blocksci_address_group_header */
