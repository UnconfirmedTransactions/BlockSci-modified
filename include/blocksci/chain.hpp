//
//  chain.hpp
//  blocksci
//
//  Created by Harry Kalodner on 9/4/17.
//
//

#ifndef chain_h
#define chain_h

#include <blocksci/chain/algorithms.hpp>
#include <blocksci/chain/block.hpp>
#include <blocksci/chain/blockchain.hpp>
#include <blocksci/chain/input_pointer.hpp>
#include <blocksci/chain/input.hpp>
#include <blocksci/chain/output_pointer.hpp>
#include <blocksci/chain/output.hpp>
#include <blocksci/chain/transaction.hpp>
#include <blocksci/chain/transaction_range.hpp>

#endif /* chain_h */
