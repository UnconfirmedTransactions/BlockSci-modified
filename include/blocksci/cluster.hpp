//
//  cluster.h
//  blocksci
//
//  Created by Harry Kalodner on 4/4/18.
//

#ifndef cluster_group_header_hpp
#define cluster_group_header_hpp

#include <blocksci/cluster/cluster.hpp>
#include <blocksci/cluster/cluster_manager.hpp>

#endif /* cluster_group_header_hpp */
