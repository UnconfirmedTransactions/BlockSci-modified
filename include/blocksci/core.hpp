//
//  core.hpp
//  blocksci
//
//  Created by Harry Kalodner on 4/22/18.
//
//

#ifndef core_group_header_h
#define core_group_header_h

#include <blocksci/core/bitcoin_uint256.hpp>
#include <blocksci/core/script_data.hpp>
#include <blocksci/core/address_types.hpp>
#include <blocksci/core/inout.hpp>
#include <blocksci/core/raw_transaction.hpp>
#include <blocksci/core/raw_block.hpp>


#endif /* core_group_header_h */
