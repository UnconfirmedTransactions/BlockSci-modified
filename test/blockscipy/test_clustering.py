import blocksci
from util import sorted_tx_list


def test_clustering_default_heuristic(chain, tmpdir_factory):
    """Tests that we can run create_clustering with path and chain only"""
    blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering_default_heuristic")), chain
    )


def test_clustering_proxy_heuristic(chain, tmpdir_factory):
    """Tests that we can run create_clustering with a proxy heuristic"""

    # output index 0
    heuristic = blocksci.heuristics.change.ChangeHeuristic(
        blocksci.Tx._self_proxy.outputs.where(lambda o: o.index == 0)
    )
    cm = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering_proxy_heuristic_0")),
        chain,
        heuristic=heuristic,
    )
    for tx in chain.blocks.txes:
        if tx.input_count > 0 and not blocksci.heuristics.is_coinjoin(tx):
            cluster = cm.cluster_with_address(tx.inputs[0].address)
            addresses = cluster.addresses.to_list()
            assert tx.outputs[0].address in addresses

    # output index 1
    heuristic = blocksci.heuristics.change.ChangeHeuristic(
        blocksci.Tx._self_proxy.outputs.where(lambda o: o.index == 1)
    )
    cm = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering_proxy_heuristic_1")),
        chain,
        heuristic=heuristic,
    )
    for tx in chain.blocks.txes:
        if (
            tx.input_count > 0
            and tx.output_count > 1
            and not blocksci.heuristics.is_coinjoin(tx)
        ):
            cluster = cm.cluster_with_address(tx.inputs[0].address)
            addresses = cluster.addresses.to_list()
            assert tx.outputs[1].address in addresses


def test_clustering_no_change(chain, json_data, regtest, tmpdir_factory):
    cm = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering")),
        chain,
        heuristic=blocksci.heuristics.change.none,
    )
    cluster = cm.cluster_with_address(
        chain.address_from_string(json_data["merge-addr-1"])
    )

    assert 3 == len(cluster.addresses.to_list())
    assert 3 == cluster.address_count()

    assert (
        chain.address_from_string(json_data["merge-addr-1"])
        in cluster.addresses.to_list()
    )
    assert (
        chain.address_from_string(json_data["merge-addr-2"])
        in cluster.addresses.to_list()
    )
    assert (
        chain.address_from_string(json_data["merge-addr-3"])
        in cluster.addresses.to_list()
    )

    assert cluster.index >= 0

    print(cluster.addresses.to_list(), file=regtest)
    print(cluster.balance(), file=regtest)
    print(cluster.balance(100), file=regtest)
    print(cluster.count_of_type(blocksci.address_type.pubkeyhash), file=regtest)
    print(cluster.count_of_type(blocksci.address_type.scripthash), file=regtest)
    print(cluster.count_of_type(blocksci.address_type.witness_pubkeyhash), file=regtest)
    print(cluster.count_of_type(blocksci.address_type.witness_scripthash), file=regtest)
    print(cluster.count_of_type(blocksci.address_type.witness_taproothash), file=regtest)
    print(sorted_tx_list(cluster.input_txes()), file=regtest)
    print(cluster.input_txes_count(), file=regtest)
    print(cluster.inputs().to_list(), file=regtest)
    print(sorted_tx_list(cluster.txes()), file=regtest)
    print(sorted_tx_list(cluster.output_txes()), file=regtest)
    print(cluster.output_txes_count(), file=regtest)
    print(cluster.outputs().to_list(), file=regtest)
    print(cluster.address_count(), file=regtest)
    print(cluster.txes(), file=regtest)
    print(cluster.type_equiv_size, file=regtest)

    for tx in chain.blocks.txes:
        if tx.input_count > 1 and not blocksci.heuristics.is_coinjoin(tx):
            cluster = cm.cluster_with_address(tx.inputs[0].address)
            addresses = cluster.addresses.to_list()
            for i in range(tx.input_count):
                assert tx.inputs[i].address in addresses

    cluster_regtest(chain, json_data, regtest, cm)


def test_clustering_with_change(chain, json_data, tmpdir_factory, regtest):
    heuristics = [
        blocksci.heuristics.change.peeling_chain.unique_change,
        blocksci.heuristics.change.optimal_change.unique_change,
        blocksci.heuristics.change.address_type.unique_change,
        blocksci.heuristics.change.locktime.unique_change,
        blocksci.heuristics.change.address_reuse.unique_change,
        blocksci.heuristics.change.client_change_address_behavior.unique_change,
        blocksci.heuristics.change.legacy.unique_change,
        blocksci.heuristics.change.none,
    ]

    for f in heuristics:
        cm = blocksci.cluster.ClusterManager.create_clustering(
            str(tmpdir_factory.mktemp("clustering")), chain, heuristic=f
        )
        cluster = cm.cluster_with_address(
            chain.address_from_string(json_data["merge-addr-1"])
        )

        assert 3 <= len(cluster.addresses.to_list())

        assert (
            chain.address_from_string(json_data["merge-addr-1"])
            in cluster.addresses.to_list()
        )
        assert (
            chain.address_from_string(json_data["merge-addr-2"])
            in cluster.addresses.to_list()
        )
        assert (
            chain.address_from_string(json_data["merge-addr-3"])
            in cluster.addresses.to_list()
        )

        assert cluster.index >= 0

        cluster_regtest(chain, json_data, regtest, cm)


def test_clustering_composability(chain, tmpdir_factory):
    nofunc = blocksci.heuristics.change.none
    compfunc = (
        blocksci.heuristics.change.legacy - blocksci.heuristics.change.legacy
    ).unique_change

    cm1 = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering")), chain, heuristic=nofunc
    )
    cm2 = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering")), chain, heuristic=compfunc
    )

    for cl in cm1.clusters():
        if cl.address_count() > 0:
            a = cl.addresses.to_list()[0]
            other_cluster = cm2.cluster_with_address(a)
            assert cl.address_count() == other_cluster.address_count()
            assert set(cl.addresses.to_list()) == set(other_cluster.addresses.to_list())


def test_clustering_ignore_coinjoin(chain, json_data, tmpdir_factory, regtest):
    addresses = (
        chain.tx_with_hash(json_data["simple-coinjoin-tx"])
        .inputs.map(lambda i: i.address)
        .to_list()
    )

    cm = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering")),
        chain,
        heuristic=blocksci.heuristics.change.none,
        ignore_coinjoin=True,
    )
    cluster = cm.cluster_with_address(addresses[0])
    cluster_addresses = cluster.addresses.to_list()
    assert 1 == len(cluster)

    for addr in addresses[1:]:
        assert addr not in cluster_addresses

    # Normal clustering should still work as expected
    cluster = cm.cluster_with_address(
        chain.address_from_string(json_data["merge-addr-1"])
    )
    assert 3 <= len(cluster.addresses.to_list())

    assert (
        chain.address_from_string(json_data["merge-addr-1"])
        in cluster.addresses.to_list()
    )
    assert (
        chain.address_from_string(json_data["merge-addr-2"])
        in cluster.addresses.to_list()
    )
    assert (
        chain.address_from_string(json_data["merge-addr-3"])
        in cluster.addresses.to_list()
    )

    cluster_regtest(chain, json_data, regtest, cm)


def test_clustering_cluster_coinjoin(chain, json_data, tmpdir_factory, regtest):
    addresses = (
        chain.tx_with_hash(json_data["simple-coinjoin-tx"])
        .inputs.map(lambda i: i.address)
        .to_list()
    )

    cm = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("clustering")),
        chain,
        heuristic=blocksci.heuristics.change.none,
        ignore_coinjoin=False,
    )
    cluster = cm.cluster_with_address(addresses[0])
    cluster_addresses = cluster.addresses.to_list()
    assert 1 < len(cluster)

    for addr in addresses:
        assert addr in cluster_addresses

    # Normal clustering should still work as expected
    cluster = cm.cluster_with_address(
        chain.address_from_string(json_data["merge-addr-1"])
    )
    assert 3 <= len(cluster.addresses.to_list())

    assert (
        chain.address_from_string(json_data["merge-addr-1"])
        in cluster.addresses.to_list()
    )
    assert (
        chain.address_from_string(json_data["merge-addr-2"])
        in cluster.addresses.to_list()
    )
    assert (
        chain.address_from_string(json_data["merge-addr-3"])
        in cluster.addresses.to_list()
    )

    cluster_regtest(chain, json_data, regtest, cm)


def cluster_regtest(chain, json_data, regtest, cm):
    ids = [
        "address-p2pkh-spend-1",
        "address-p2sh-spend-2",
        "addr-2-in-2-out",
        "addr-peeling-chain",
        "addr-merge-0",
        "addr-merge-2",
        "merge-addr-1",
        "merge-addr-2",
    ]
    addresses = [chain.address_from_string(json_data[x]) for x in ids]
    for addr in addresses:
        cluster = cm.cluster_with_address(addr)
        print(
            sorted(cluster.addresses.to_list(), key=lambda x: x.address_num),
            file=regtest,
        )


def test_tagged_address(chain, tmpdir_factory):
    cm = blocksci.cluster.ClusterManager.create_clustering(
        str(tmpdir_factory.mktemp("tagged-address-test")),
        chain,
        heuristic=blocksci.heuristics.change.none
    )
    address = chain[-1].txes[0].outputs[0].address
    cluster = cm.cluster_with_address(address)
    tags = {address: "test-tag"}
    assert cluster.tagged_addresses(tags).size == 1
    assert cluster.tagged_addresses(tags).to_list()[0].address == address
    assert cluster.tagged_addresses(tags).to_list()[0].tag == "test-tag"
