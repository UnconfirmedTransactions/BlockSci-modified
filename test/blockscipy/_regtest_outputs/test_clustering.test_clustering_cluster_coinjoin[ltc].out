[PubkeyHashAddress(mrPRMELJcGsbKmuFiGwUgNCX9Wo4iSy44T)]
[ScriptHashAddress(QVAf2inEFpzWnz6tTxLLYSXYJR2KzA4kEE)]
[PubkeyHashAddress(mkZ9Np6JVYyHC35zEdfUXdM6rwrMKPpTZa)]
[PubkeyHashAddress(msexSMRCUhqikgCA5FjA1DjMXyv4BWhSbb)]
[PubkeyHashAddress(muJKkgEiFx6Ws5vcRRtNyaSM96D95X8kRj)]
[PubkeyHashAddress(mo9cbu8M5YhhLRNMn65tbdrn124uNQDxsL)]
[PubkeyHashAddress(muPiDgt2BB8omxzFTVEpATbKU5cpUPo6tr), PubkeyHashAddress(mokMSp4xbaU7rvpZAZyxmonZWQoFnEiPny), PubkeyHashAddress(mhULRhVhUKD23jbao92nUE97DDDFkCVpec)]
[PubkeyHashAddress(muPiDgt2BB8omxzFTVEpATbKU5cpUPo6tr), PubkeyHashAddress(mokMSp4xbaU7rvpZAZyxmonZWQoFnEiPny), PubkeyHashAddress(mhULRhVhUKD23jbao92nUE97DDDFkCVpec)]
